ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp

```diff
@@ -122,33 +159,104 @@ size_t TensorLayout::compute_page_size_bytes(const Size& page_size) const {
+    // TODO: Alignment before sharding can describe padding along non-height or width dims
+    // However, this is not compatible with logical sharding and we just leave it as nullopt
+    const int alignment_rank = static_cast<int>(alignment_.size());
+    TT_FATAL(alignment_rank <= 2, "Alignment {} must be rank 2 or less to compute physical shard shape", alignment_);
+
+    auto physical_shard_height = CMAKE_UNIQUE_NAMESPACE::round_up(logical_shard_shape.height(), alignment_[-2]);
+    auto physical_shard_width = CMAKE_UNIQUE_NAMESPACE::round_up(logical_shard_shape.width(), alignment_[-1]);
+    return Size{physical_shard_height, physical_shard_width};
+}
+
 Size TensorLayout::compute_physical_shape(const ttnn::SimpleShape& shape) const {
     const int rank = static_cast<int>(shape.rank());
     const int alignment_rank = static_cast<int>(alignment_.size());
-    const int max_rank = std::max(rank, alignment_rank);
+
     size_t width = 1;
     size_t height = 1;
 
-    // Iterate dims in reverse order and ensure alignment
-    // Even tensor of rank 0 or 1 must be aligned (to Tile / Page / Shard)
-    for (int i = -1; i >= -max_rank; --i) {
-        auto& dim = i == -1 ? width : height;
-        if(i >= -rank) {
+    if (memory_config_.shard_spec.has_value() and memory_config_.shard_spec.value().mode == ShardMode::LOGICAL) {
+        // Iterate dims in reverse order
+        for (int i = -1; i >= -rank; --i) {
+            auto& dim = i == -1 ? width : height;
             dim *= shape[i];
         }
 
-        // Align the current dimension if alignment is available
-        if (i >= -alignment_rank) {
-            dim = CMAKE_UNIQUE_NAMESPACE::round_up(dim, alignment_[i]);
+        auto compute_default_logical_shard_shape = [&shape, &rank]() {
+            size_t shard_width = 1;
+            size_t shard_height = 1;
+
+            if (rank >= 1) {
+                shard_width = shape[-1];
+            }
+            if (rank >= 2) {
+                shard_height = shape[-2];
+            }
+
+            return Size{shard_height, shard_width};
+        };
+
+        const auto& logical_shard_shape = memory_config_.shard_spec.has_value() ? Size(memory_config_.shard_spec.value().shape) : compute_default_logical_shard_shape();
+        const auto& physical_shard_shape = compute_physical_shard_shape(logical_shard_shape);
+
+        auto get_physical_size = [](auto original_size, auto logical_shard_size, auto physical_shard_size, auto alignment) -> uint32_t {
+            if (logical_shard_size == 0) {
+                return 0;
+            }
+            // If we always pad to full shards, then return:
+            // auto num_shards = tt::div_up(original_size, logical_shard_size);
+            // return (uint32_t) physical_shard_size * num_shards;
+
+            // But host physical data is only padded up to nearest alignment (and no padding in between shards)
+            // So last shard is only up to nearest alignment
```

Comment:
This comment with commented out code seems a bit odd, can we rephrase it?
For example: 
```
All shards except for the last one are padded to full shard size, but the last one is just padded up to the nearest alignment.
```

Reply:
I left other code in as an option. Because we don't actually do the conversion yet, it can go either way.